source/extensions/quic/connection_id_generator/quic_lb/quic_lb.cc

```diff
@@ -0,0 +1,353 @@
+  auto new_cid = tls_slot_->encoder_.MaybeReplaceConnectionId(original, version);
+  if (new_cid.has_value()) {
+    return appendRoutingId(new_cid.value(), original);
+  }
+
+  return new_cid;
+}
+
+uint8_t QuicLbConnectionIdGenerator::ConnectionIdLength(uint8_t first_byte) const {
+  return tls_slot_->encoder_.ConnectionIdLength(first_byte) + kRoutingInfoSize;
+}
+
+absl::optional<quic::QuicConnectionId>
+QuicLbConnectionIdGenerator::appendRoutingId(quic::QuicConnectionId& new_connection_id,
+                                             const quic::QuicConnectionId& old_connection_id) {
+  uint8_t buffer[quic::kQuicMaxConnectionIdWithLengthPrefixLength];
+
+  const uint16_t new_length = new_connection_id.length() + kRoutingInfoSize;
+  if (new_length > sizeof(buffer)) {
+    IS_ENVOY_BUG("Invalid encoder configuration led to connection id being too long");
+    return {};
+  }
+
+  if (old_connection_id.length() < kRoutingInfoSize) {
+    IS_ENVOY_BUG("Unexpected very short connection id");
+    return {};
+  }
+
+  memcpy(buffer, new_connection_id.data(), new_connection_id.length());
+
+  uint8_t* routing_info_destination = buffer + new_connection_id.length();
+  const uint8_t* routing_info_source = reinterpret_cast<const uint8_t*>(old_connection_id.data()) +
+                                       old_connection_id.length() - kRoutingInfoSize;
+  memcpy(routing_info_destination, routing_info_source, kRoutingInfoSize);
+
+  return quic::QuicConnectionId(absl::Span<const uint8_t>(buffer, new_length));
+}
+
+QuicLbConnectionIdGenerator::ThreadLocalData::ThreadLocalData(
+    const envoy::extensions::quic::connection_id_generator::quic_lb::v3::Config& config,
+    absl::string_view server_id)
+    : encoder_(*quic::QuicRandom::GetInstance(), nullptr /* visitor */,
+               config.self_encode_length()),
+      unsafe_unencrypted_testing_mode_(config.unsafe_unencrypted_testing_mode()),
+      nonce_length_bytes_(config.nonce_length_bytes()), server_id_(server_id) {}
+
+absl::StatusOr<std::shared_ptr<QuicLbConnectionIdGenerator::ThreadLocalData>>
+QuicLbConnectionIdGenerator::ThreadLocalData::create(
+    const envoy::extensions::quic::connection_id_generator::quic_lb::v3::Config& config,
+    const absl::string_view server_id) {
+
+  std::shared_ptr<QuicLbConnectionIdGenerator::ThreadLocalData> ret(
+      new QuicLbConnectionIdGenerator::ThreadLocalData(config, server_id));
+
+  return ret;
+}
+
+absl::Status
+QuicLbConnectionIdGenerator::ThreadLocalData::updateKeyAndVersion(const std::string& key,
+                                                                  uint8_t version) {
+  std::optional<quic::LoadBalancerConfig> lb_config;
+  if (unsafe_unencrypted_testing_mode_) {
+    lb_config = quic::LoadBalancerConfig::CreateUnencrypted(version, server_id_.length(),
+                                                            nonce_length_bytes_);
+  } else {
+    lb_config =
```

Comment:
When you do thread encoding, it's important to append the thread ID to the server ID and shorten the nonce accordingly (which means you have to verify the nonce isn't now too short). Otherwise different threads will generate the same cipher text and potentially threaten the thread protection key. I'll update Section 6.2 of the spec to make this explicit.

Reply:
The entire area of thread protection key has been omitted in this PR. I'll add a TODO in the code. As this PR currently exists, the thread ID is unencoded/unprotected/plaintext in the CID, which adds a small amount of linkability.

FWIW, the default CID generator in envoy is much worse: keeps the first 4 bytes of the original CID, and copies that into all subsequent CIDs, which adds a huge amount of linkability. And the rest of the new CID is just a hash of the previous CID.

So even with a small amount of linkability from the thread id, this is still substantially less trackable than the default.